### ISSUE TITLE
database abstraction

### ISSUE THREAD

**Issue by stevenvachon** [chiel/omni-old#28]
To remove the dependency on MongoDB. People may want to use RethinkDB or PostgreSQL, etc.

**Comment by chiel** [comment]
I suppose that will be somewhat difficult though - the nice thing about mongo and friends is that they're schemaless, so you don't have to worry about migrations and stuff like that, which usually really causes headaches.

I agree that people might want to use other database types though, so it's an interesting topic to discuss - if you have any suggestions on how to solve this in a sane way, I'm open to your thoughts!

**Comment by stevenvachon** [comment]
Something like https://github.com/grncdr/node-any-db

Perhaps that lib could be rewritten with a plugin-like architecture.

**Comment by chiel** [comment]
Well, that lib appears to only be for sql databases accprding to the docs: `postgres`, `mysql`, `sqlite3` and `mssql` - which means it avoids the whole headache of having to deal with nosql vs sql databases, since it just caters to one type of database. Sure, those listed differ to an extent, but in the end they have similar feature sets.

Another potential issue is that at the moment I am using `mongoose` to ensure that the database entities get created correctly, and that each field gets coerced into the correct type. If the choice to support multiple types of databases is made, this functionality will also have to exist or be created.

Aside from this, one downside of using database abstraction layers is that often they are quite generic and don't really leverage each database type's full potential. This can sometimes lead to fairly significant performance loss.

Mind you, I'm not saying database abstraction is off the table, but I think if we go down that route it should probably be solved at the application level, probably in a module's model.

At the moment there is actually still too much stuff happening in the router which actually belongs in the model, so I do intend to strip this out as soon as there's a v0.1.0. It will hopefully not take too long before I get there.

Cheers for the input so far, and apologies for the wall of text!

**Comment by stevenvachon** [comment]
Np, very informative. I created [this issue](https://github.com/grncdr/node-any-db/issues/68) for any-db. Perhaps omni could also benefit from a plugin architecture for swapping features. It could remove permanent dependencies on redis and mongoose.

**Comment by chiel** [comment]
Yes, my goal with this project is to be able to create a custom cms in a matter of hours, with completely customised modules and such, that's why I've been focussing a lot on making stuff fairly generic, also when it comes to generating the mongoose schemas, routing and development workflow.

Stuff like that is definitely candidate for being made more generic/pluggable, I guess you can almost think of the current state as more of a proof of concept than anything. It's definitely nearing a state where it's usable but once that's done, it's time to break it open and refactor since I'm fairly certain there's stuff that could be solved in a better way.
